pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,353 @@
+package ftp
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/textproto"
+	"os"
+	"path"
+	"time"
+
+	"github.com/jlaffaye/ftp"
+)
+
+// Conn struct embeds the *ftp.ServerConn returned by ftp server on successful connection.
+type Conn struct {
+	*ftp.ServerConn
+}
+
+// Retr wraps the ftp retrieve method to return a ftpResponse interface type.
+func (c *Conn) Retr(filepath string) (ftpResponse, error) {
+	return c.ServerConn.Retr(filepath)
+}
+
+func (c *Conn) RetrFrom(filepath string, offset uint64) (ftpResponse, error) {
+	return c.ServerConn.RetrFrom(filepath, offset)
+}
+
+var (
+	errEmptyFilename            = errors.New("filename cannot be empty")
+	errEmptyPath                = errors.New("file/directory path cannot be empty")
+	errEmptyDirectory           = errors.New("directory name cannot be empty")
+	errInvalidArg               = errors.New("invalid filename/directory")
+	transferCompleteError       = &textproto.Error{Code: 226, Msg: "Transfer complete."}
+	directoryAlreadyExistsError = &textproto.Error{Code: 550, Msg: "Create directory operation failed."}
```

Comment:
```suggestion
	errTransferComplete       = &textproto.Error{Code: textproto.StatusClosingDataConnection, Msg: "Transfer complete."}
	errDirectoryAlreadyExists = &textproto.Error{Code: 550, Msg: "Create directory operation failed."}
```

Reply:
Instead of Code, I need to check the message because a similar code is returned in various cases. 

for e.g : 
226 is Closing data connection. Requested file action successful (for example, file transfer or file abort).
550 is Requested action not taken. File unavailable (e.g., file not found, no access).

Reply:
It's a pity then, maybe something you cannot fix.

I'm worried about the fact each FTP server may reply different message based on their implementation.

Maybe, @jlaffaye (hi :wave: ) who maintain the ftp lib might help, or suggest something here.

Reply:
any suggestion here @puellanivis ? thanks

Reply:
It’s a sticky spot… I think a way to work around it might be to work backwards, and try and stat the file with `GetEntry` if this errors, move up the chain of dirs. (Eventually, at worst you should hit `/` and that should work. If it doesn’t then err permission out.)

At that point, you can then proceed down the directory tree making the directories as you come back up the call stack. Eventually, it either errors and you bubble that up, or you’re on the last, and you return success.

Reply:
I noticed something and suggested it here

https://github.com/gofr-dev/gofr/pull/871#discussion_r1701275839